Non-wrapping mode for better Transformers compatibility

evgri243 opened the pull request:
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs change / refactoring / dependency upgrade

## Motivation and Context / Related issue

The primary goal of this PR is to introduce a **Non-wrapping mode** (controller-based) to Opacus. This feature allows for the computation of per-sample gradients without wrapping the original `nn.Module` in a `GradSampleModule` subclass. 

This change addresses critical compatibility issues with third-party libraries—most notably HuggingFace `Transformers` and `Accelerate`—which often rely on strict module hierarchies or perform type checks that are disrupted by standard Opacus wrappers.

**Key Technical Changes:**
- **`GradSampleHooks` Controller**: Introduced a new architecture where gradient sampling hooks are managed by a standalone `GradSampleHooks` object instead of being embedded in a `GradSampleModule` wrapper.
- **`PrivacyEngine` Updates**: Modified `make_private` and `make_private_with_epsilon` to support a `wrap_model` parameter (defaults to `True` for backward compatibility). When `wrap_model=False`, Opacus attaches hooks to the original model and returns it along with a `GradSampleHooks` controller.
- **Enhanced Loss Compatibility**: Updated `DPLossFastGradientClipping` to handle custom loss implementations that lack a `reduction` attribute (a common occurrence in the `transformers` library). The implementation now explicitly sets the expected reduction if missing and restores the previous state after the backward pass.
- **HuggingFace Trainer Integration**: Verified the new mode via successful small-scale training runs using the HuggingFace `Trainer` with Fast Gradient Clipping. A new example (`examples/huggingface_trainer.py`) is included to demonstrate this integration.
- **Architectural Cleanup**: Refactored `gsm_base.py` to provide a unified `AbstractGradSampleHooks` base class, improving the robustness of hook attachment, attribute cleanup, and validation logic across all sampling modes.

## How Has This Been Tested (if it applies)

- **New Unit Tests**: 
- **Integration Testing**: Successfully performed training with the HuggingFace `Trainer` using the new non-wrapping mode for both next-token prediction and SFT/DPO.

## Checklist

- [x] The documentation is up-to-date with the changes I made (including a new tutorial and updated READMEs).
- [x] I have read the **CONTRIBUTING** document and completed the CLA.
- [x] All tests passed, and additional code has been covered with new tests.

iden-kalemaj commented:
That's a good point... let me know how it's looking once you work on it.

evgri243 committed:
Refactor grad sample modules to separate hooks logic from module functionality. Introduce `GradSampleHooks` to handle per-sample gradient computation independently of `nn.Module` structure.

evgri243 committed:
Refactor privacy engine to add support for hooks-based gradient sampling without model wrapping. Introduce optional `wrap_model` argument and update gradient sampling modes. Rename and consolidate hooks-related classes.

evgri243 committed:
Add tutorial for training with non-wrapping mode (`wrap_model=False`) and update references in documentation and metadata.

evgri243 committed:
Expand hooks-based gradient sampling documentation and improve argument details across modules.

evgri243 committed:
Add `wrap_model` support across tests and modules for flexible gradient sampling. Integrate new hooks-based classes and multi-device handling scenarios. Update tests for FP16, BF16, and mixed precision.

evgri243 committed:
Add HuggingFace Trainer integration example with Differential Privacy (DP) support.

evgri243 commented:
Ok. Calling it "as limited as possible" is an overstatement, but I've drastically changed the design and did my best to avoid unnecessary changes

iden-kalemaj commented on opacus/grad_sample/grad_sample_module_fast_gradient_clipping.py:
```diff
@@ -40,31 +40,32 @@
                 [], device=grad_sample.device, dtype=grad_sample.dtype
             )
         else:
-            param._norm_sample = torch.zeros(
```
what is the reason for removing this? this was added for low-precision and mixed-precision functionality.

iden-kalemaj commented on opacus/tests/mixed_precision_test.py:
```diff
@@ -73,27 +74,34 @@ def _get_training_components(
         # Make the model private with the specified precision
         if grad_sample_mode in ["hooks", "functorch", "ew"]:
-            model, optimizer, dataloader = privacy_engine.make_private(
+            hooks_or_wrapper, optimizer, dataloader = privacy_engine.make_private(
```
nit: call it hooks_or_gs_module

evgri243 commented on examples/huggingface_trainer.py:
```diff
@@ -0,0 +1,552 @@
+"""
+HuggingFace Trainer with Differential Privacy support.
+This example demonstrates a robust integration of Opacus with HuggingFace Trainer,
+following the original implementation used in federated research.
```
Removed and paraphrased. It slipped...

evgri243 committed:
Attend to comments

evgri243 committed:
Further simplifying changes

evgri243 commented:
I  addressing your cutting all unnecessary changes surgically. Should be out first half next week.

evgri243 committed:
Ensure safe restoration of state in gradient computation and loss reduction methods to stabilize gpu mixed-precision tests.

evgri243 committed:
Remove unused `peft` dependency from `dev_requirements.txt`.

iden-kalemaj reviewed (approved):
Thank you again @evgri243 for contributing this heavy lift functionality. I've left a few minor comments, but this looks good to me now. 

Just to check, were you able to run the multi-GPU unit tests in a multi-gpu environment? Github CI does not support the multi-gpu tests for now. If not, I can run them.

evgri243 committed:
Attend to PR comments

iden-kalemaj reviewed (commented):
hi @evgri243, did you intend to submit this PR_MERGE_PLAN?

evgri243 committed:
Fix test stability and improve hook handling

- Safer remove_hooks(): no ValueError when hooks already removed
- Initialize grad_accumulation_hook = None in all GradSampleModule subclasses
- Add warning in cleanup() when hook removal fails (still cleans attributes)
- Remove redundant cleanup() call in hooks test
- Add prepare_module() as new main API, keep wrap_model() as backward compat alias
- Use prepare_module import in privacy_engine instead of aliased wrap_model
- Fix criterion.reduction = 'none' in DPLossFastGradientClipping
- Clean up imports in adaptive_clipping_utils

evgri243 commented:
Ofc not... a short moment of shame.

Back to changes. I squashed all renaming back and forth and returned to what it used to be + some changes that popped up during test stabilization.

iden-kalemaj commented:
@evgri243 no worries, can you rebase to main and resolve conflicts after we landed some changes from your team?

let's try to land your PR soon after that

evgri243 commented:
I guess I forgot to push. I did that

iden-kalemaj commented:
@evgri243 for some reason the system is failing to import the PR internally. I'll check back on this tomorrow morning.

iden-kalemaj commented:
@evgri243  can you try rebasing and seeing if there are any conflicts, we are unable to import the PR internally which is usually due to merge conflicts.

evgri243 commented:
Rebased and pushed. Let's see.

PS. I am so slow as I am not happy with the resulting diamond inheritance between module and hooks. But so far I fail to think of a better way.

If we get it merged, may I ask you to trigger a build release with all the changes we submitted. This is exactly what was required to make our trainer work. We will rebase it on upstream Opacus and test it thoroughly in the training pipeline before we make further decisions. What do you think?

iden-kalemaj commented:
@evgri243 that sounds good since we're due for another version release. The current solution looks elegant enough :)

Please see the lint failure and to speed things up please test with all 3 linters [here](https://github.com/meta-pytorch/opacus/blob/main/CONTRIBUTING.md#our-development-process).

evgri243 committed:
Fix flake issue: refactor test assertions for multi-device parameter handling in GradSampleModuleFastGradientClippingEmbeddingLayerTest

evgri243 commented:
Interestingly. It was something old. Fixed it

evgri243 commented:
See massive failures. Attending to them today.

evgri243 committed:
Restore DPTensorFastGradientClipping arithmetic operations

evgri243 commented:
CI/CD seems to be finally green. We rebased our code on top of this branch and previously merged changes -- seems working, but we will continue thorough testing.

iden-kalemaj commented:
@evgri243 i did not get to check the initial failures, can you let me know why the tests were failing? Just want to make sure that we are not duplicating any of the code in #805

evgri243 commented:
The errors were just because I removed all the changes with arithmetics. This PR had them originally and then removed in favor of an independent PR. Seeing that, rebase silently undid them as well
